zmq.hpp

```diff
@@ -615,8 +620,340 @@ inline void swap(context_t &a, context_t &b) ZMQ_NOTHROW {
     a.swap(b);
 }
 
+#ifdef ZMQ_CPP11
+struct send_result
```

Comment:
While this would restrict this to C++17, send_result and recv_result could be defined as a std::optional<size_t>. While the use of operator*/-> have UB in case no value is set, misuses would be more obvious, and probably trigger an assertion in typical standard library implementations.

Reply:
The initial thought was that additional members could be added to the struct in a backwards compatible fashion, but since it is highly unlikely that the libzmq API will change I agree that that is a more robust and better solution to use `std::optional`.

We could still support C++11 by returning a "implementation defined" type (only pre-C++17) that emulates the basic operations of optional for trivial types, just `operator*`, `operator->`, `has_value()` and `operator bool`. For C++17 we just use `std::optional`.

Reply:
The existing libzmq API functions will probably never be changed. Theoretically new functions could be added that would be used in the existing cppzmq functions, but this appears rather unlikely to me as well. 

If you would like to add such a minimal std::optional drop-in for older standards, that would be great!